core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy

```diff
@@ -115,23 +115,44 @@ class ConnectorConfiguratorTomcat {
     } else if (connectors.size() < 1) {
       return null
     } else {
-      throw new IllegalStateException("Unexpected server.xml format - one http connector expected at most")
+      throw new IllegalStateException("Unexpected server.xml format - one ajp connector expected at most")
     }
   }
 
   private boolean isSecured(Node connector) {
     return connector.@secure.toString().toLowerCase().trim() == "true"
   }
 
-  private void createNewConnector(Map<String, Object> attributes) {
-    server.Service.each { service ->
-      service.appendNode ("Connector", attributes)
+
+  private void createNewConnector(Node element) {
+    server.Service.each { Node service ->
+      service.appendNode(element, element.attributes(), element.value())
     }
   }
 
-  private void updateExistingConnector(Node Connector, Map<String, Object> attributes) {
-    attributes.each { attribute ->
-      Connector.@"${attribute.key}" = attribute.value
+  private void updateExistingConnector(Node connector, Node newConnector) {
+    // update attributes
+    newConnector.attributes().each { attribute ->
+      connector.@"${attribute.key}" = attribute.value
+    }
+
+    // update nodes
+    newConnector.each { Node newSubelement ->
+      String UpgradeProtocol = ConnectorUpgradeProtocolTomcat.ELEMENT_NAME
+
+      if (newSubelement.name() == UpgradeProtocol) {
+        if (connector.find { it.name() == UpgradeProtocol } == null) {
```

Comment:
`== null` is superfluous

Reply:
taking back, it is not, sorry for the noise

Reply:
> Ok, did the tests passed with all the last changes? If so, I am ready to merge if you don't plan to do any additional changes.

Give some time to verify it once again